File: flat_panel_velocity.m

%%velocity at (x,y) induced by a flat source panel from (x1,y1) to (x2,y2) of
%strength lambda_1
function velocity=flat_panel_velocity(x1,y1,x2,y2,x,y,lambda_1)
%flat panel end point 1
p1=[x1 y1 0];
%flat panel end point 2
p2=[x2 y2 0];
%point of interested velocity 
p=[x y 0];
r1=p-p1;
r2=p-p2;
r1_mag=sqrt(r1(1,1)^2+r1(1,2)^2);
r2_mag=sqrt(r2(1,1)^2+r2(1,2)^2);
p1p2=p2-p1;
p1p2_mag=sqrt(p1p2(1,1)^2+p1p2(1,2)^2);
t_hat=1/p1p2_mag*(p2-p1);
z_hat=[0 0 1];
r1r2_cross=cross(r1,r2);
alpha=atan2(dot(r1r2_cross,z_hat),dot(r1,r2));
n_hat=cross(z_hat,t_hat);
velocity=lambda_1/2/pi*(log(r1_mag/r2_mag)*t_hat+alpha*n_hat);

end

File: flat_panel_velocity.py

"""flat_panel_velocity is to calculate the velocity at desired position induced by a flat 
   source panel from pos1 to pos2 of strength lambda_1
   or to calculate the velocity at desired postion induced by a flat vortex panel 
   from pos1 to pos2 of strength gamma_1
...moduleauthor: Qingquan Wang<qwang252@wisc.edu> Yang Lou<lou9@wisc.edu>"""
import numpy as np
import math
def source_panel_velocity(pos1,pos2,desd_pos,lambda_1):
  # set position 1 in numpy array 
    pos1=np.array([pos1[0],pos1[1],0])
  # set position 2 in numpy array
    pos2=np.array([pos2[0],pos2[1],0])
  # set desired position in numpy array
    desd_pos=np.array([desd_pos[0],desd_pos[1],0])
    r1=desd_pos-pos1
    r2=desd_pos-pos2   
    r1_mag=np.linalg.norm(r1)
    r2_mag=np.linalg.norm(r2)
    p1p2=pos2-pos1
    p1p2_mag=np.linalg.norm(p1p2)
    t_hat=1/p1p2_mag*(p1p2)
    z_hat=np.array([0,0,1])
    r1r2_cross=np.cross(r1,r2)
    alpha=math.atan2(np.dot(r1r2_cross,z_hat),np.dot(r1,r2))
    n_hat=np.cross(z_hat,t_hat)
    
   # governing velocity equation for source panel method
    source_velocity=lambda_1/2/math.pi*(np.log(r1_mag/r2_mag)*t_hat+alpha*n_hat)
    return source_velocity 

def vortex_panel_velocity(pos1,pos2,desd_pos,gamma_1):
   # set position 1 in numpy array 
    pos1=np.array([pos1[0],pos1[1],0])
   # set position 2 in numpy array
    pos2=np.array([pos2[0],pos2[1],0])
   # set desired position in numpy array
    desd_pos=np.array([desd_pos[0],desd_pos[1],0])
    r1=desd_pos-pos1
    r2=desd_pos-pos2
    r1_mag=np.linalg.norm(r1)
    r2_mag=np.linalg.norm(r2)
    p1p2=pos2-pos1
    p1p2_mag=np.linalg.norm(p1p2)
    t_hat=1/p1p2_mag*(p1p2)
    z_hat=np.array([0,0,1])
    r1r2_cross=np.cross(r1,r2)
    alpha=math.atan2(np.dot(r1r2_cross,z_hat),np.dot(r1,r2))
    n_hat=np.cross(z_hat,t_hat)

   # governing velocity equation for vortex panel method
    vortex_velocity=gamma_1/2/math.pi*(-np.log(r1_mag/r2_mag)*n_hat+alpha*t_hat)
    return vortex_velocity
